src/System.Text.Json/docs/ReferenceHandling_spec.md

```diff
@@ -0,0 +1,433 @@
+                    }
+                ]
+            }            
+        }
+    }";
+
+    var options = new JsonSerializerOptions
+    {
+        ReferenceHandling = ReferenceHandling.Preserve
+    };
+
+    Employee angela = JsonSerializer.Deserialize<Employee>(json, options);
+    Console.WriteLine(object.ReferenceEquals(angela, angela.Manager.Subordinates[0])); //prints: true.
+}
+```
+
+On Newtonsoft's Json.Net:
+```cs
+public static void ReadJsonWithPreservedReferences(){
+    string json = 
+    @"{
+        ""$id"": ""1"",
+        ""Name"": ""Angela"",
+        ""Manager"": {
+            ""$id"": ""2"",
+            ""Name"": ""Bob"",
+            ""Subordinates"": {
+                ""$id"": ""3"",
+                ""$values"": [
+                    { 
+                        ""$ref"": ""1"" 
+                    }
+                ]
+            }            
+        }
+    }";
+
+    var options = new JsonSerializerSettings
+    {
+        MetadataPropertyHanding = MetadataPropertyHandling.Default //Json.Net reads metadata by default, just setting the option for ilustrative purposes.
+    };
+
+    Employee angela = JsonConvert.DeserializeObject<Employee>(json, settings);
+    Console.WriteLine(object.ReferenceEquals(angela, angela.Manager.Subordinates[0])); //prints: true.
+}
+```
+
+# Ground rules
+## Unsupported types
+Basically, any type that uses a `EnumerableConverter` and tries to be preserved will throw.
+
+* **Immutable types**: i.e: `ImmutableList` and `ImmutableDictionary`
+* **System.Array**
+
+Aside from above, the Deserializer will throw when a CLR value type is passed as preserved when `ReferenceHandling.Preserve` is set. 
+
+As a rule of thumb, we should throw on all cases where the payload contains metadata that is impossible to create with the `JsonSerializer`, however, this conflicts with feature parity in Json.Net; those scenarios are described below.
+
+## Reference objects ($ref)
+
+* Regular property **before** `$ref`.
+  * **Json.Net**: `$ref` is ignored if a regular property is previously found in the object.
+  * **S.T.Json**: Throw - Reference objects cannot contain other properties.
+
+```json
+{
+    "$id": "1",
+    "Name": "Angela",
+    "ReportsTo": {
+        "Name": "Bob",
+        "$ref": "1"
+    }
+}
+```
+
+* Regular property **after** `$ref`.
+  * **Json.Net**: Throw - Additional content found in JSON reference object.
+  * **S.T.Json**: Throw - Reference objects cannot contain other properties.
+
+```json
+{
+    "$id": "1",
+    "Name": "Angela",
+    "ReportsTo":{
+        "$ref": "1",
+        "Name": "Angela" 
+    }
+}
+```
+ 
+* Metadata property **before** `$ref`:
+  * **Json.Net**: `$id` is disregarded and the reference is set.
+  * **S.T.Json**: Throw - Reference objects cannot contain other properties.
+```json
+{
+    "$id": "1",
+    "Name": "Angela",
+    "ReportsTo": {
+        "$id": "2",
+        "$ref": "1"
+    }
+}
+```
+
+* Metadata property **after** `$ref`:
```

Comment:
Add rule about `JsonExtensionData` where metadata won't get added to it ever (even if we ignore it with a setting).

Reply:
I though we agreed that metadata will be added to `JsonExtensionData` if we choose to ignore it.

Reply:
> I though we agreed that metadata will be added to `JsonExtensionData` if we choose to ignore it.

Is ignoring metadata the default behavior on deserialization? I am fine with that. Please document it as such.